### PR TITLE
[SPARK-35785][SS] Cleanup support for RocksDB instance

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
Add the functionality of cleaning up files of old versions for the RocksDB instance and RocksDBFileManager.

### Why are the changes needed?
Part of the implementation of RocksDB state store.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New UT added.

### CODE REVIEW

**Comment by viirya** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:225]: Is this the change for review in this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:207]: Do you mean, if SST file F was last used in version V, then it won't be used in version V+2 or later?

In other words, a SST file F can be only used in continuous versions. It won't be used in V, not used in V+1, and then used again in V+2.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:223]: What the second case for? When it will happen?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:273]: maxVersionPresent -> maxUsedVersion?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:286]: version -> versionFile.

Actually, maybe `s"Error deleting version file $versionFile for version $version"`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:289]: Maybe we need to reduce the files failed to delete?

**Review by viirya — COMMENTED** [review]
If `deleteOldVersions` is the only change needed for review, then it looks okay. Just a few minor comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:225]: That's right.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:27]: This will fail to pass scalastyle now.

**Review by HeartSaVioR — COMMENTED** [review]
First round. Need to look into test cases a bit more.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:206]: That's right. The file can not be shared with skipping versions. If a file used in V and not used in V+1, the checkpoint of V+1 should already create new files for all the KVs included in the original file.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:204]: Thanks, done in the next commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:222]: Yep, that's right.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:263]: Thanks, done in the next commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:272]: Ah thanks, done in the next commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:285]: Make sense, done in the next commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:288]: Agree, done in the next commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:390]: Yes, the `localTempDir` is only used for unzip files.

**Comment by HeartSaVioR** [comment]
Could you please rebase this again? Looks like we have some "stale" changes here, like RocksDBLoader.

You'll probably need to "cherry-pick" commits into the new branch, and do "hard reset" against this branch, and do force push.

**Comment by xuanyuanking** [comment]
@HeartSaVioR Addressed done, the redundant code should be introduced by my old commits. Thanks for reminding me!

**Review by HeartSaVioR — APPROVED** [review]
+1 if test passes.

As usual I'd like to make sure @viirya also approves this - I'll wait for a couple of days before merging this.

**Comment by viirya** [comment]
I'll take another look tomorrow. Thanks!

**Comment by xuanyuanking** [comment]
Thanks again for the help of @HeartSaVioR and @viirya.
The test failure is not related to the changes. Let me retrigger.

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:256]: When will we call this method?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:276]: This test seems not related to clean up change here? Looks like more related to RocksDB instance PR.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:351]: Hm, what this test is used for? Each `RocksDB` in each thread uses the same remote root dir, won't they conflict?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:256]: It will be called in the `RocksDBStateStoreProvider.doMaintenace`. I'll submit the state store provider PR (the last one) today.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:256]: okay.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (xuanyuanking) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:276]: Ah yea, this is the test for rollback.
Actually the original plan is expose `rollback` and `cleanup` in this PR. It should be a mistake for the last PR, I introduced the `rollback` without tests.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:276]: okay

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (xuanyuanking) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:351]: This is to simulate the multi-thread scenario of updating and cleaning old versions. It will not conflict since we call commit for each update thread and the version get updated for each commits.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:351]: Hmm, will it happens? I think `RocksDB` is not thread-safe, and each state task only has one `RocksDB` instance. They should update and clean old versions individually as they are for different state store.

Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:351]: This looks to be more likely simulating the case multiple streaming queries with same checkpoint run concurrently. 

SST files shouldn't conflict as we make the file name be unique, and for metadata files we use `overwriteIfPossible = true`, so won't throw error if the file already exists.

Comment (xuanyuanking) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:351]: Yea, I think the purpose of this test is to make sure no error thrown and the result is correct in the end.
After taking a further look, there's a small issue is that `exception` never used. I'll confirm it separately.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:351]: @xuanyuanking and me discussed this test offline. Seems there is something wrong with `exception` usage. It doesn't look completely correct. @xuanyuanking will address it by fixing it or deleting the test later in a follow-up.

**Review by viirya — APPROVED** [review]
lgtm

**Comment by viirya** [comment]
Thanks @xuanyuanking for working on this and @HeartSaVioR for the review! Merging to master.

**Comment by xuanyuanking** [comment]
Great thanks for the help! @viirya @HeartSaVioR 
I'll update the rest PR and submit the last one of RocksDBStateStoreProvider today.

**Comment by viirya** [comment]
Ah, sorry, I forgot this should be in branch-3.2 too. @xuanyuanking Can you submit a PR for 3.2?

**Comment by xuanyuanking** [comment]
Sure. Let me do it now.

**Comment by dongjoon-hyun** [comment]
Hi, All.
After this PR, Jenkins seems to fail twice consecutively. Could you take a look at that please, @xuanyuanking , @HeartSaVioR , @viirya ?
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-3.2/2711/testReport/org.apache.spark.sql.execution.streaming.state/RocksDBSuite/ensure_that_concurrent_update_and_cleanup_consistent_versions/history/

cc @gengliangwang

**Comment by dongjoon-hyun** [comment]
It seems that https://issues.apache.org/jira/browse/SPARK-35993 is filed already by @attilapiros

**Comment by viirya** [comment]
Thanks @dongjoon-hyun! Let me ignore the test first to unblock others. @xuanyuanking will address (fix or delete) the test later.

**Pull request closed without merging**
